### ISSUE TITLE
Distributing Comparisons

### ISSUE THREAD

**Issue by kirkgrubbs1** [MrOlm/inStrain#98]
Hey, I have 350 metagenomic samples that I essentially want to perform an all v all comparison on. I have already been able to produce profiles for each. Up until the final "compare" step things are pretty straightforward to distribute across an HPC environment.

For the inStrain compare portion, I was wondering if it would be valid to divide up the comparisons into batches and then recombine results at the end. Also, does the directionality of a comparison matter (like BLAST)?

Thus far I have written a wrapper that enumerates all pairwise comparisons, splits them into batches based on whatever value the user wants, removes duplicates within each batch (I'm assuming inStrain compare will do the all v all comparison for each batch and, if it matters, do the comparisons in each direction), and then starts each batch as its own job.

These were the basic questions I have come across thus far, but I'll probably have more when combining results at the end.

**Comment by MrOlm** [comment]
Hello,

A couple of things. First, directionality does not matter with inStrain; you'll get the same results either way.

Second, right now, inStrain only does pairwise comparisons. There's no ability to select specific comparisons. Also by default it will only do the comparisons in one direction, and will not compare samples against themselves.

In my experience, once you get up to 100 samples or so, the pairwise comparisons start taking a long time. This is just the nature of pair-wise comparisons. I've approached this problem by launching 1 inStrain compare job **for each genome to be compared**, and **only comparing samples with the genome robustly detected.** So of your 350 metagenomes, if genome X is detected in 10 samples, launch an inStrain compare job that focuses only on genome X and only the 10 inStrain profiles with that genome.

Does this make sense? 

Best,
Matt

**Comment by kirkgrubbs1** [comment]
Yes that makes sense. Thank you for the clarification and suggestions.

**Comment by kirkgrubbs1** [comment]
after thinking through things more closely, isn't you suggestion basically what database mode does, but as one job rather than an individual job for each genome?

**Comment by MrOlm** [comment]
Yep that's exactly right

**Comment by kirkgrubbs1** [comment]
I have taken your advice and only performed comparisons with genomes that show up in at least two profiles. This has drastically cut down on comparisons and, in-turn, run time. Unfortunately, I'm getting several new errors in logs and stdout. I was wondering if you could illuminate their meanings a bit. They are as follows:

Error 1
> ..:: Failures ::..
> The following scaffolds were in the .stb file given, but not the original .fasta file used for profiling. They will not be considered in genomeLevel operations:
> NODE_2362_length_27726_cov_4.103791 (intended for genome _p912-22001701533083.103.fa)

This goes on to list 630,817 additional contigs.

Error 2
> ..:: inStrain compare Step 3. Auxiliary processing ::..
> ***************************************************
>                 
> 22-01-04 19:05:25 DEBUG    FAILURE StbError NODE_2362_length_27726_cov_4.103791 p912-22001701533083.103.fa no_length will not be considered as part of the genome

This goes on to list 630,869 additional contigs.

I have gone back and checked and double-checked to make sure I am using the same fasta and stb files as in the profiles. So, I'm kinda at a loss for what is going on here. 

All that said, I still get output, and the rest of the comparison run seems to complete successfully.

**Comment by MrOlm** [comment]
Glad that you were able to successfully cut run-time!

I assume you ran with database mode, which would explain all these errors. In database-mode, inStrain removes from consideration all scaffolds that don't have at least some number of reads mapped. This really reduces run-time. The .stb file you're giving it, though does have these scaffolds in it. This confuses inStrain and makes it throw and error.

So in short, I designed these errors in a time before "database_mode", and they're not relevant when using database_mode. I'll fix this behavior in the next inStrain update.

Best,
Matt

**Comment by kirkgrubbs1** [comment]
hmm, I'm not actually using database mode, at least I don't think so, an example of the command format I'm using follows:

> inStrain compare --genome 1.91.fa -o 1.91 -s scaffold_to_bin.stb -p 6 -i 1.IS 2.IS 3.IS

(actual labels changed because of data sensitivity)

**Comment by MrOlm** [comment]
Sorry I should have clarified- I meant running database_mode on the "profile" step, not the "compare" step

**Comment by kirkgrubbs1** [comment]
Oh, I gotcha, sounds good. Thank you for all your help.

**Comment by liamfitzstevens** [comment]
Hi @MrOlm ,

Just wanted to let you know that the `DEBUG    FAILURE StbError Part1_Sample7__maxbin2__LowProb_051__CONTIG1 Part1_Sample7__maxbin2__LowProb_051.fna no_length will not be considered as part of the genome` behavior as originally mentioned by @kirkgrubbs1 is still present in `inStrain compare Step 3. Auxiliary processing`. Sample log files for both `profile` and `compare` are attached for your reference.

Best,
Liam

[profile.log](https://github.com/MrOlm/inStrain/files/11768311/profile.log)
[compare.log](https://github.com/MrOlm/inStrain/files/11768312/compare.log)

**Comment by MrOlm** [comment]
Hi @liamfitzstevens - thank you for letting me know. I will fix this in the next inStrain update.

Best,
Matt
